diff --git a/lsp/gcl_dl.lsp b/lsp/gcl_dl.lsp
index 6f6b35f44..aa5adcc5f 100644
--- a/lsp/gcl_dl.lsp
+++ b/lsp/gcl_dl.lsp
@@ -3,9 +3,16 @@
 
 (export '(mdlsym mdl lib-name))
 
+(defvar *dll-extension* #+darwin ".dylib" #+cygwin ".dll" #-(or darwin cygwin) ".so")
+
 (defun lib-name (p)
-  (if (or (string= p "") (string= p "libc") (string= p "libm")) "" 
-    (string-concatenate #+darwin "/usr/lib/system/" p #+darwin ".dylib" #+cygwin ".dll" #-(or darwin cygwin) ".so")));FIXME
+  (cond ((or (string= p "") (string= p "libc") (string= p "libm"))
+         "")
+        #+darwin ; "/usr/lib/system/" is macOS-specific
+        ((zerop (string-match "libsystem_" p))
+         (string-concatenate "/usr/lib/system/" p *dll-extension*))
+        (t
+         (string-concatenate #+darwin "/usr/lib/" p *dll-extension*))))
 
 (defun mdl (n p vad)
   (let* ((sym (mdlsym n (lib-name p)))
